### ISSUE TITLE
Policy issue: TLSv1.3 makes upgrade without program modification impossible

### ISSUE THREAD

**Issue by levitte** [openssl/openssl#5944]
TLSv1.3 is different from TLS versions up to v1.2.  This means that programs that want to use it must be modified, whether they explicitly asked for TLSv1.3, or implicitly by just asking for "maximum available version".

This means that there's a potential that upgrading OpenSSL from 1.1.0 to 1.1.1 will causes program crashes (the regressions reported in #5661 are a strong indication).

We need a manner of detecting if a program built against an older version of the library is executed, and if that is the case, only allow the maximum TLS version to be 1.2.

**Comment by t8m** [comment]
I do not count the case where the application really expects to negotiate only TLSv1.2 and nothing else.

Of course if you really come to conclusion that limiting to TLSv1.2 for applications that do not know about TLSv1.3 is necessary. Then the easiest solution is to set max_protocol version to TLSv1.2 by default and require everything that wants to support TLSv1.3 to set it to TLSv1.3 explicitly.

**Comment by levitte** [comment]
Well, as #5661 has shown, *everything* that uses TLS, is built against 1.1.0 and doesn't explicitely specify TLSv1.2 as their maximum version goes belly up.

**Comment by t8m** [comment]
But these are tests which have much different expectations from what things like "simple" TLS client which does not care about the protocol details have.

**Comment by t8m** [comment]
And do the tests actually start to work if you just recompile them against 1.1.1 without any code change? That would actually indicate ABI break.

**Comment by levitte** [comment]
Good questions.  I'm looking further into that

**Comment by richsalz** [comment]
Yeah, I'm leaning against this issue.  If the library falls over because it found an unknown version, that's really bad. It's a bug.  So I am not sure what this is trying to address.

**Comment by richsalz** [comment]
And also, I *strongly* agree that the tests are not a good indicator.  In fact, they're a *bad* indicator.

**Comment by mattcaswell** [comment]
Some applications built against 1.1.0 will work correctly when using that library, but if someone upgrades the library to 1.1.1 they will suddenly stop working if TLSv1.3 is negotiated - because TLSv1.3 behaves differently to TLSv1.2. If an application is built with TLSv1.2 assumptions that don't hold in TLSv1.3 then this is where problems can occur. This won't impact the majority of applications, but some might be.

**Comment by levitte** [comment]
> And also, I *strongly* agree that the tests are not a good indicator.  In fact, they're a *bad* indicator.

In that case, I'm looking forward to good indicators that a library upgrade won't cause things to fall over as soon as OpenSSL is upgraded.  All *I* know are our test programs, so it's not so much about good vs bad indicators, it's more about any available indicator against none at all.

Quite frankly, I think it's more important to figure out *why* things fall over (some might be justifiable, I dunno) and fix things to avoid that when possible (I'm counting that's the case for the majority of the issues that are found) than ignoring this and having it come back and bite us in form of an angry comunity.

**Comment by mattcaswell** [comment]
> And also, I strongly agree that the tests are not a good indicator. In fact, they're a bad indicator.

I think they are a good indicator of the types of problems we might encounter. They're not a good indicator of how likely those problems are to occur.

**Comment by levitte** [comment]
Thanks, that does make more sense

**Comment by kaduk** [comment]
Just to go on the record: it's still unclear to me how big the issue really is, and whether that would justify a high-impact change like #5945 .

**Comment by levitte** [comment]
The fact that 1.1.0 TLS specific test programs (as well as `apps/openssl s_server` and `apps/openssl s_client` all fall over is an indicator that the issue exists.  I would be strongly surprised if those were the only ones affected...  or if they are, that gets me worried another way, because that would indicate how far away our test programs are from real world application regarding what they try to do.

**Comment by t8m** [comment]
But why does the simple combination of s_server / s_client from 1.1.0 fall over when run with 1.1.1?

**Comment by mattcaswell** [comment]
> But why does the simple combination of s_server / s_client from 1.1.0 fall over when run with 1.1.1?

Does it? I just tried it and it worked for me.

**Comment by levitte** [comment]
I'm sorry, that might be a an exageration.  The 1.1.0 tests in group 70 are more or less all going belly up when run with a 1.1.1 shared library (see #5661).  So it's not exactly the most default use of `s_client` and `s_server`...

**Comment by dot-asm** [comment]
> The 1.1.0 tests in group 70 are more or less all going belly up when run with a 1.1.1 shared library (see #5661)

This is of course because the TLSProxy is examining actual protocol flow and is expecting specific sequence of packets and even specific packet contents. But this should not be motivation to punish the common application. Because common application doesn't care about packet sequence or content. So that as far as TLSProxy goes, if setting max_protocol doesn't cut it, then it would be more appropriate to suppress the tests if library is newer than application. More appropriate than to deny common application to use new protocol that is.

Of course it still might happen that application would get confused by alternative sequence of callbacks. But in such case it would probably be more appropriate to find a way to limit specifically these applications to 1.2...

In the essence it all boils down to following question. If you need to recompile your application in order to use TLS 1.3, then what do you need 1.1.1 for? I mean to have TLS 1.3 in a *minor* release. As it could as well be major...

**Comment by dot-asm** [comment]
> what about test/sslapitest?

As it stands now I get couple of errors in 70 range and test_ssl_new. How do I reproduce your sslapitest problem? Straight #5947?

**Comment by levitte** [comment]
> As it stands now I get couple of errors in 70 range and test_ssl_new. How do I reproduce your sslapitest problem? Straight #5947?

Yup.  Against a master without this PR.

**Comment by dot-asm** [comment]
> Straight #5947?

Looking at **first** failure alone. It's all about session id, something that is not available in 1.3. So wouldn't it be reasonable to assume that *if* application sets any session id callbacks or corresponding control flags, it wouldn't know how to deal with 1.3 and automatically set a cap on negotiated TLS version to 1.2?

**Comment by levitte** [comment]
We've [discussed quite a bit on openssl-project](https://mta.openssl.org/pipermail/openssl-project/2018-April/000570.html), and it's clear that the issue wasn't as black and white as I thought (I'm backing off from #5945 as a result).

I do think that #5947 needs to happen.  The commits it reverts are the wrong path.
I do think that we need to discuss the edge cases where programs that get relinked with libssl 1.1.1 may fall over just because TLSv1.3 came in and took over, and hopefully solve these cases in a somewhat more smooth manner than just bailing out.
